File: api/registration_example.py
"""
Exemplo de lógica de registo contra a API FinScan
Base URL: https://finscan-production.up.railway.app
Rota de registo: POST /api/v1/auth/register
"""

import requests
import json
from typing import Optional, Dict, Any


class FinScanRegistration:
    """Cliente para registo e autenticação na FinScan API"""
    
    BASE_URL = "https://finscan-production.up.railway.app"
    REGISTER_ENDPOINT = "/api/v1/auth/register"
    LOGIN_ENDPOINT = "/api/v1/auth/login"
    
    def __init__(self, base_url: Optional[str] = None):
        """
        Inicializa o cliente
        
        Args:
            base_url: URL base da API (default: https://finscan-production.up.railway.app)
        """
        self.base_url = base_url or self.BASE_URL
    
    def register(self, username: str, email: str, password: str, role: str = "user") -> Dict[str, Any]:
        """
        Regista um novo utilizador
        
        Args:
            username: Nome de utilizador único
            email: Email válido do utilizador
            password: Password da conta
            role: Papel do utilizador ('user' ou 'admin', default: 'user')
        
        Returns:
            Dict contendo resposta da API
            
        Raises:
            ValueError: Se os dados forem inválidos
            requests.exceptions.RequestException: Se o pedido falhar
        
        Exemplo:
            >>> client = FinScanRegistration()
            >>> response = client.register(
            ...     username="joao_silva",
            ...     email="joao@example.com",
            ...     password="senha_segura123",
            ...     role="user"
            ... )
            >>> print(response)
            {'id': 1, 'username': 'joao_silva', 'email': 'joao@example.com', 'is_active': True, 'role': 'user'}
        """
        
        # Validações básicas
        if not username or not username.strip():
            raise ValueError("Username não pode estar vazio")
        
        if not email or not email.strip():
            raise ValueError("Email não pode estar vazio")
        
        if not password or not password.strip():
            raise ValueError("Password não pode estar vazia")
        
        if role not in ["user", "admin"]:
            raise ValueError("Role deve ser 'user' ou 'admin'")
        
        # Validar email
        if "@" not in email or "." not in email.split("@")[1]:
            raise ValueError("Email inválido")
        
        # Preparar dados
        payload = {
            "username": username.strip(),
            "email": email.strip(),
            "password": password,
            "role": role
        }
        
        # Fazer pedido
        url = f"{self.base_url}{self.REGISTER_ENDPOINT}"
        
        try:
            response = requests.post(
                url,
                json=payload,
                timeout=10
            )
            
            # Processar resposta
            if response.status_code == 201:
                # Registo bem-sucedido
                return response.json()
            
            elif response.status_code == 400:
                # Erro de validação
                error_detail = response.json().get("detail", "Erro desconhecido")
                raise ValueError(f"Erro de validação: {error_detail}")
            
            elif response.status_code == 422:
                # Erro de validação de schema
                errors = response.json().get("detail", [])
                error_msg = "; ".join([
                    f"{e.get('loc', ['unknown'])[1]}: {e.get('msg', 'erro desconhecido')}"
                    for e in errors
                ])
                raise ValueError(f"Erro de schema: {error_msg}")
            
            else:
                raise Exception(f"Erro da API (status {response.status_code}): {response.text}")
        
        except requests.exceptions.Timeout:
            raise Exception("Timeout ao conectar à API")
        except requests.exceptions.ConnectionError:
            raise Exception("Erro de conexão com a API")
    
    def login(self, username: str, password: str) -> Dict[str, str]:
        """
        Autentica um utilizador e retorna um token JWT
        
        Args:
            username: Nome de utilizador
            password: Password da conta
        
        Returns:
            Dict contendo o token de acesso
            
        Exemplo:
            >>> client = FinScanRegistration()
            >>> token_response = client.login("joao_silva", "senha_segura123")
            >>> print(token_response)
            {'access_token': 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9...', 'token_type': 'bearer'}
        """
        
        if not username or not password:
            raise ValueError("Username e password são obrigatórios")
        
        # Usar OAuth2PasswordRequestForm format
        data = {
            "username": username,
            "password": password
        }
        
        url = f"{self.base_url}{self.LOGIN_ENDPOINT}"
        
        try:
            response = requests.post(
                url,
                data=data,  # OAuth2 usa form data, não JSON
                timeout=10
            )
            
            if response.status_code == 200:
                return response.json()
            
            elif response.status_code == 401:
                raise ValueError("Username ou password incorretos")
            
            else:
                raise Exception(f"Erro da API (status {response.status_code}): {response.text}")
        
        except requests.exceptions.Timeout:
            raise Exception("Timeout ao conectar à API")
        except requests.exceptions.ConnectionError:
            raise Exception("Erro de conexão com a API")


def main():
    """Exemplo de utilização"""
    
    client = FinScanRegistration()
    
    # Exemplo 1: Registo bem-sucedido
    print("=" * 60)
    print("EXEMPLO 1: Registo de novo utilizador")
    print("=" * 60)
    
    try:
        user_data = {
            "username": "joao_silva",
            "email": "joao@example.com",
            "password": "senha_super_segura123",
            "role": "user"
        }
        
        print(f"\nTentando registar com os seguintes dados:")
        print(json.dumps(user_data, indent=2, ensure_ascii=False))
        
        response = client.register(**user_data)
        
        print(f"\n✓ Registo bem-sucedido!")
        print(f"Resposta da API:")
        print(json.dumps(response, indent=2, ensure_ascii=False))
        
    except ValueError as e:
        print(f"\n✗ Erro de validação: {e}")
    except Exception as e:
        print(f"\n✗ Erro: {e}")
    
    # Exemplo 2: Tentar registar com username duplicado
    print("\n" + "=" * 60)
    print("EXEMPLO 2: Tentativa de registo com username duplicado")
    print("=" * 60)
    
    try:
        response = client.register(
            username="joao_silva",  # Mesmo username do exemplo anterior
            email="outro_email@example.com",
            password="outra_senha123"
        )
        print(f"\n✓ Registo bem-sucedido: {response}")
        
    except ValueError as e:
        print(f"\n✗ Erro esperado: {e}")
    except Exception as e:
        print(f"\n✗ Erro: {e}")
    
    # Exemplo 3: Login
    print("\n" + "=" * 60)
    print("EXEMPLO 3: Login após registo bem-sucedido")
    print("=" * 60)
    
    try:
        token_response = client.login("joao_silva", "senha_super_segura123")
        
        print(f"\n✓ Login bem-sucedido!")
        print(f"Token de acesso recebido:")
        print(json.dumps({
            "access_token": token_response.get("access_token", "")[:50] + "...",
            "token_type": token_response.get("token_type", "")
        }, indent=2, ensure_ascii=False))
        
    except ValueError as e:
        print(f"\n✗ Erro: {e}")
    except Exception as e:
        print(f"\n✗ Erro: {e}")
    
    # Exemplo 4: Validações - Email inválido
    print("\n" + "=" * 60)
    print("EXEMPLO 4: Validação - Email inválido")
    print("=" * 60)
    
    try:
        response = client.register(
            username="novo_user",
            email="email_invalido",  # Email inválido
            password="senha123"
        )
    except ValueError as e:
        print(f"\n✓ Validação funcionou: {e}")
    except Exception as e:
        print(f"\n✗ Erro: {e}")
    
    # Exemplo 5: Validações - Password vazia
    print("\n" + "=" * 60)
    print("EXEMPLO 5: Validação - Password vazia")
    print("=" * 60)
    
    try:
        response = client.register(
            username="novo_user",
            email="novo@example.com",
            password=""  # Password vazia
        )
    except ValueError as e:
        print(f"\n✓ Validação funcionou: {e}")
    except Exception as e:
        print(f"\n✗ Erro: {e}")


if __name__ == "__main__":
    # Descomentar a linha abaixo para executar os exemplos
    # main()
    
    # Ou usar o cliente manualmente:
    client = FinScanRegistration()
    
    # Registo
    try:
        response = client.register(
            username="test_user",
            email="test@example.com",
            password="test_password123"
        )
        print("Registo bem-sucedido:", response)
    except Exception as e:
        print("Erro:", e)

File: api/registration_example.js
/**
 * Cliente FinScan Registration
 * Base URL: https://finscan-production.up.railway.app
 * Rota de registo: POST /api/v1/auth/register
 */

class FinScanRegistration {
  constructor(baseUrl = "https://finscan-production.up.railway.app") {
    this.baseUrl = baseUrl;
    this.registerEndpoint = "/api/v1/auth/register";
    this.loginEndpoint = "/api/v1/auth/login";
  }

  /**
   * Valida um email
   * @param {string} email - Email a validar
   * @returns {boolean} - True se o email é válido
   */
  validateEmail(email) {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return emailRegex.test(email);
  }

  /**
   * Regista um novo utilizador
   * @param {string} username - Nome de utilizador único
   * @param {string} email - Email válido
   * @param {string} password - Password da conta
   * @param {string} role - Papel do utilizador ('user' ou 'admin', default: 'user')
   * @returns {Promise<Object>} - Dados do utilizador criado
   * @throws {Error} - Se houver erro na validação ou na API
   *
   * @example
   * const client = new FinScanRegistration();
   * try {
   *   const user = await client.register(
   *     "joao_silva",
   *     "joao@example.com",
   *     "senha_segura123",
   *     "user"
   *   );
   *   console.log("Utilizador criado:", user);
   * } catch (error) {
   *   console.error("Erro:", error.message);
   * }
   */
  async register(username, email, password, role = "user") {
    // Validações básicas
    if (!username || !username.trim()) {
      throw new Error("Username não pode estar vazio");
    }

    if (!email || !email.trim()) {
      throw new Error("Email não pode estar vazio");
    }

    if (!password || !password.trim()) {
      throw new Error("Password não pode estar vazia");
    }

    if (!["user", "admin"].includes(role)) {
      throw new Error("Role deve ser 'user' ou 'admin'");
    }

    // Validar email
    if (!this.validateEmail(email)) {
      throw new Error("Email inválido");
    }

    // Preparar payload
    const payload = {
      username: username.trim(),
      email: email.trim(),
      password: password,
      role: role,
    };

    try {
      const url = `${this.baseUrl}${this.registerEndpoint}`;
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
      });

      const data = await response.json();

      if (response.status === 201) {
        // Registo bem-sucedido
        return data;
      } else if (response.status === 400) {
        // Erro de validação
        throw new Error(`Erro de validação: ${data.detail}`);
      } else if (response.status === 422) {
        // Erro de schema
        const errorMsg = data.detail
          .map((e) => `${e.loc[1]}: ${e.msg}`)
          .join("; ");
        throw new Error(`Erro de schema: ${errorMsg}`);
      } else {
        throw new Error(
          `Erro da API (status ${response.status}): ${data.detail || "erro desconhecido"}`
        );
      }
    } catch (error) {
      if (error instanceof Error && error.message.startsWith("Erro")) {
        throw error;
      }
      throw new Error(`Erro de conexão: ${error.message}`);
    }
  }

  /**
   * Autentica um utilizador
   * @param {string} username - Nome de utilizador
   * @param {string} password - Password
   * @returns {Promise<Object>} - Token de acesso JWT
   * @throws {Error} - Se houver erro na autenticação
   *
   * @example
   * const client = new FinScanRegistration();
   * try {
   *   const token = await client.login("joao_silva", "senha_segura123");
   *   console.log("Token:", token.access_token);
   *   // Guardar o token no localStorage
   *   localStorage.setItem("access_token", token.access_token);
   * } catch (error) {
   *   console.error("Erro de autenticação:", error.message);
   * }
   */
  async login(username, password) {
    if (!username || !password) {
      throw new Error("Username e password são obrigatórios");
    }

    try {
      const url = `${this.baseUrl}${this.loginEndpoint}`;

      // OAuth2PasswordRequestForm usa form data
      const formData = new URLSearchParams();
      formData.append("username", username);
      formData.append("password", password);

      const response = await fetch(url, {
        method: "POST",
        body: formData,
      });

      const data = await response.json();

      if (response.status === 200) {
        return data;
      } else if (response.status === 401) {
        throw new Error("Username ou password incorretos");
      } else {
        throw new Error(
          `Erro da API (status ${response.status}): ${data.detail || "erro desconhecido"}`
        );
      }
    } catch (error) {
      if (error instanceof Error && error.message.startsWith("Erro")) {
        throw error;
      }
      throw new Error(`Erro de conexão: ${error.message}`);
    }
  }
}

// ============================================
// EXEMPLOS DE UTILIZAÇÃO
// ============================================

/**
 * Exemplo 1: Registo simples
 */
async function exemploRegisto() {
  const client = new FinScanRegistration();

  try {
    console.log("Registando novo utilizador...");
    const user = await client.register(
      "joao_silva",
      "joao@example.com",
      "senha_segura123",
      "user"
    );

    console.log("✓ Registo bem-sucedido!");
    console.log("Dados do utilizador:", user);
    // {
    //   id: 1,
    //   username: "joao_silva",
    //   email: "joao@example.com",
    //   role: "user",
    //   is_active: true
    // }
  } catch (error) {
    console.error("✗ Erro:", error.message);
  }
}

/**
 * Exemplo 2: Login após registo
 */
async function exemploLogin() {
  const client = new FinScanRegistration();

  try {
    console.log("Fazendo login...");
    const token = await client.login("joao_silva", "senha_segura123");

    console.log("✓ Login bem-sucedido!");
    console.log("Token:", token.access_token);

    // Guardar token no localStorage (para frontend)
    localStorage.setItem("access_token", token.access_token);
    localStorage.setItem("token_type", token.token_type);
  } catch (error) {
    console.error("✗ Erro:", error.message);
  }
}

/**
 * Exemplo 3: Fluxo completo de registo + login
 */
async function exemploFluxoCompleto() {
  const client = new FinScanRegistration();

  try {
    // Passo 1: Registo
    console.log("Passo 1: Registando novo utilizador...");
    const user = await client.register(
      "maria_santos",
      "maria@example.com",
      "senha_super_segura123"
    );
    console.log("✓ Utilizador criado:", user.username);

    // Passo 2: Login
    console.log("\nPasso 2: Fazendo login...");
    const token = await client.login("maria_santos", "senha_super_segura123");
    console.log("✓ Login bem-sucedido!");

    // Passo 3: Usar o token para fazer requests autenticados
    console.log("\nPasso 3: Usando token para request autenticado...");
    const response = await fetch(
      "https://finscan-production.up.railway.app/api/v1/users/me",
      {
        method: "GET",
        headers: {
          Authorization: `Bearer ${token.access_token}`,
        },
      }
    );

    const userData = await response.json();
    console.log("✓ Dados do utilizador:", userData);
  } catch (error) {
    console.error("✗ Erro:", error.message);
  }
}

/**
 * Exemplo 4: Tratamento de erros
 */
async function exemploTratamentoErros() {
  const client = new FinScanRegistration();

  // Erro 1: Email inválido
  try {
    await client.register("user1", "email_invalido", "senha123");
  } catch (error) {
    console.log("Erro esperado:", error.message);
    // "Email inválido"
  }

  // Erro 2: Username duplicado (já registado)
  try {
    await client.register(
      "joao_silva",
      "novo_email@example.com",
      "senha123"
    );
  } catch (error) {
    console.log("Erro esperado:", error.message);
    // "Erro de validação: Username or email already exists"
  }

  // Erro 3: Password incorreta no login
  try {
    await client.login("joao_silva", "password_errada");
  } catch (error) {
    console.log("Erro esperado:", error.message);
    // "Username ou password incorretos"
  }
}

// Exportar para usar em módulos
if (typeof module !== "undefined" && module.exports) {
  module.exports = FinScanRegistration;
}
